spec/features/time_slots_spec.rb

```diff
@@ -1,71 +1,94 @@
-  end
+      it "displays the timeslot properly on the time slots page" do
+        visit '/time_slots'
+        expect_time_slot_to_display_properly(@slot, time_slot_row(@slot.location.id, Date.today).find('li.timeslot'))
+      end
 
-  xit "displays the timeslot properly on the shifts page" do
-    visit '/shifts'
-  end
-end
+      # Note that the part of time_slot in the past is not open
+      it "displays the timeslot properly on the shifts schedule page" do 
+        visit shifts_path
+        expect_time_slot_to_display_properly(@slot, shift_schedule_row(@slot.location.id, Date.today).find('li.bar_open'))
+      end
 
+      # Note that when new loc_groups are created, the roles are NOT updated 
+      # with the newly-generated permissions, therefore this new timeslot would
+      # not be displayed on the timeslots page
+      it "can update timeslot" do 
+        c = create(:calendar); loc = create(:location)
+        visit edit_time_slot_path(@slot)
+        select c.name, from: "Calendar"
+        fill_in_date("time_slot_start_date", @slot.start+2.day)
+        select loc.name, from: "Location"
+        click_on "Save Changes"
+        expect_flash_notice "Successfully updated timeslot"
+        expect{@slot.reload}.to change(@slot, :start)
+        expect(@slot.calendar).to eq(c)
+        expect(@slot.location).to eq(loc)
+      end
 
-#
-## Common Actions
-#
+      # Alternatively, one can use selenium to handle alert
+      it "can destroy timeslot" do 
+        visit time_slot_path(@slot)
+        click_on "Destroy"
+        # If using selenium driver
+        # page.driver.browser.switch_to.alert.accept
+        expect_flash_notice "Successfully destroyed timeslot"
+        expect{TimeSlot.find(@slot.id)}.to raise_error(ActiveRecord::RecordNotFound)
+      end
 
-def app_setup
-  @app_config = create(:app_config)
-  @department = create(:department)
-  @loc_group = build(:loc_group, department: @department)
-  @location = create(:location, loc_group: @loc_group)
-end
+    end
+  end
 
-def create_timeslot
-  visit '/time_slots/new'
-  within("#new_time_slot") do
-    # fill_in DATE, :with => TOMORROW
-    # fill_in time_slot_start_time_4i, :with => "10" #10am
-    # ...more time
+  context "For repeating time_slots" do 
+    # Need JavaScript for the deletion submenu to work
+    context "Different types of deletion", js: true do
+      before :each do 
+        @re = create(:repeating_time_slots, location_ids: [@location.id], calendar: @calendar)
+        # getting the second timeslot in the series
+        @slot = @re.time_slots.order(:start).second
+        @count = TimeSlot.count
+        visit edit_time_slot_path(@slot)
+        click_on "Destroy this time slot"
+      end
+      it "can delete one single instance of timeslot" do 
+        click_on "Just this time slot"
+        expect_flash_notice "Successfully destroyed timeslot"
+        expect{TimeSlot.find(@slot.id)}.to raise_error(ActiveRecord::RecordNotFound)
```

Comment:
Measuring change is usually more problematic since you need to wrap your operations in a block and then also need to worry about webpage loading latency etc...it sometimes led to strange failures so I didn't use it unless it is necessary.